File: 2023/compass/script/fix_reaction.R
options(warn = -1)
suppressMessages({
    library(argparser)
    library(tidyverse)
    library(data.table)
})
options(warn=1)

# args --
argv <- arg_parser('')
argv <- add_argument(argv, "--reactions", help = "compass reactions")
argv <- add_argument(argv, "--metadata", help = "compass reaction metadata")
argv <- add_argument(argv, "--outdir", help = "output dir")
argv <- add_argument(argv, "--name", help = "name")
argv <- parse_args(argv)

argv$metadata <- ifelse(is.na(argv$metadata), "/SGRNJ06/randd/USER/wangjingshen/script/compass/data/reaction_metadata.csv", argv$metadata)
argv$outdir <- ifelse(is.na(argv$outdir), "./", argv$outdir)
if(!dir.exists(argv$outdir)){
    dir.create(argv$outdir)
}
argv$name <- ifelse(is.na(argv$name), "", paste0("_", argv$name))

# read --
res = read.table(argv$reactions, sep = "\t", header = T, row.names = 1)
ref <- fread(argv$metadata, encoding = 'UTF-8')

# function from compass
function_get_reaction_consistencies <- function(compass_reaction_penalties, min_range=1e-3){
    df = -log(compass_reaction_penalties + 1)
    df = df[apply(res,1,max) - apply(res,1,min) >= min_range,]
    df = df - min(df)
    return(df)
}
res <- function_get_reaction_consistencies(res)

res_ref = ref[match(sapply(row.names(res), function(x){substring(x, 1, nchar(x) - 4)}), ref$reaction_no_direction),]
row.names(res) <- paste0(row.names(res), "___", res_ref$subsystem, "___", res_ref$reaction_name)

# out
write.table(res, paste0(argv$outdir, "/reaction_consistencies", argv$name, ".tsv"), quote = F, sep="\t")

File: projects/2023/compass/script/fix_reaction.py
import argparse
import os
import subprocess
import pandas as pd
import numpy as np


def get_reaction_consistencies(compass_reaction_penalties, min_range=1e-3):
    """
    Converts the raw penalties outputs of compass into scores per reactions where higher numbers indicate more activity
    """
    df = -np.log(compass_reaction_penalties + 1)
    df = df[df.max(axis=1) - df.min(axis=1) >= min_range]   # filter reactions
    df = df - df.min().min()
    return df

def run(reactions, metadata, outdir, name):
    reaction_penalties = pd.read_csv(reactions, sep="\t", index_col = 0)
    reaction_consistencies = get_reaction_consistencies(reaction_penalties)
    # metadata
    reaction_metadata = pd.read_csv(metadata, index_col = 0)

    test =reaction_consistencies.reset_index().reset_index().iloc[:,0:2]
    test['reaction'] = test['index'].apply(lambda x: x[:-4])
    test1 = test.merge(reaction_metadata, how='left', left_on='reaction', right_index=True, validate='m:1')
    # check method 1
    #reaction_consistencies.index.to_list() == test1['index'].to_list()
    # check method 2
    #sum([(reaction_consistencies.index[i]) == (test1['index'][i]) for i in range(6490)])
    reaction_consistencies.index = reaction_consistencies.index + "___" + test1.subsystem + "___" + test1.reaction_name    
    reaction_consistencies.to_csv(f'{outdir}/reaction_consistencies{name}.tsv')


def main():
    parsers = argparse.ArgumentParser()
    parsers.add_argument('--reactions', help='reactions', required=True)
    parsers.add_argument('--metadata', help='metadata')
    parsers.add_argument('--outdir', help='outdir')
    parsers.add_argument('--name', help='name')

    args = parsers.parse_args()
    if args.metadata == None:
        args.metadata="/SGRNJ06/randd/USER/wangjingshen/script/compass/data/reaction_metadata.csv"
    if args.outdir == None:
        args.outdir="./"
    if not os.path.exists(args.outdir):
        os.system(f'mkdir -p {args.outdir}')
        #cmd1 = f'mkdir -p {outdir}'
        #subprocess.check_call(cmd1, shell = True)
    if args.name == None:
        args.name=""
    if args.name!='':
        args.name = "_" + args.name
    run(args.reactions, args.metadata, args.outdir, args.name)

if __name__ == '__main__':
    main()
